Prose/ProseLib/Sources/SharedModels/JID.swift

```diff
@@ -0,0 +1,169 @@
+//
+//  JID.swift
+//  Prose
+//
+//  Created by Rémi Bardon on 07/06/2022.
+//
+
+import Foundation
+import Parsing
+
+/// JIDs, as defined in [XEP-0029](https://xmpp.org/extensions/xep-0029.html).
+///
+/// Grammar:
+/// ```grammar
+/// <JID> ::= [<node>"@"]<domain>["/"<resource>]
+/// <node> ::= <conforming-char>[<conforming-char>]*
+/// <domain> ::= <hname>["."<hname>]*
+/// <resource> ::= <any-char>[<any-char>]*
+/// <hname> ::= <let>|<dig>[[<let>|<dig>|"-"]*<let>|<dig>]
+/// <let> ::= [a-z] | [A-Z]
+/// <dig> ::= [0-9]
+/// <conforming-char> ::= #x21 | [#x23-#x25] | [#x28-#x2E] |
+///                       [#x30-#x39] | #x3B | #x3D | #x3F |
+///                       [#x41-#x7E] | [#x80-#xD7FF] |
+///                       [#xE000-#xFFFD] | [#x10000-#x10FFFF]
+/// <any-char> ::= [#x20-#xD7FF] | [#xE000-#xFFFD] |
+///                [#x10000-#x10FFFF]
+/// ```
+public struct JID: Hashable {
+    public let node: String?
+    public let domain: String
+    public let resource: String?
+}
+
+extension JID: CustomStringConvertible {
+    public var description: String {
+        (try? jidParserPrinter.print(self)).map(String.init) ?? "<error>"
+    }
+}
+
+extension JID: ExpressibleByStringLiteral {
+    public init(stringLiteral value: String) {
+        self = try! jidParserPrinter.parse(value)
```

Comment:
First, I can see you commented on this `try!`, I'm realizing it could have been a `if let {} else { fatalError/assertionFailure }` to have a nicer error message.

Yeah, I wrote a few tests (not extensive, jut to be sure everything looks OK), which saved me from comitting versions that were compiling but toally wrong!

The benefits I see are:
- Obviously great Swift integration.
- Parsing and printing supported.
- Understandable error description, but as we won't show it to the user, I didn't try to make it readable, and it should not be taken into consideration.
- Easier to migrate later from an existing Swift `struct`, rather than a `String`.

The drawbacks:
- Overkill for a simple JID. Makes sense for long files, but this seems like a steam machine.
- A new dependency, but as stated in my first commit, we will have this dependency anyway if we decide to use `URLRouting` from Point-Free.
- Longer compile time: with a lot of generics, type inference can take a long time, or even not succeed. I try to add some types when they can really help the compiler, but I try not to add too much to not hurst readability. I've had quite a lot of issues with the current generics system recently, I'm hoping it will be fixed in the next months/years.
- Duplicate code
- Nebulous code (I tought it would be readable, but it isn't)

It's mainly drawbacks, I agree that it wasn't necessary.

Reply:
Thanks, for pointing out the benefits and drawbacks. Since you seem to agree that there are mainly drawbacks is there a reason why we want to keep the _parsing code_?

We'd have mostly the same benefits when the parsing is done in the core library.
- We'd still have a struct. We could even create another struct on the Swift side that wraps the struct returned from Rust. Or add extensions on the Rust struct, e.g. `ExpressibleByStringLiteral`.
- Printing a jid can be trivially done by string concatenation and wouldn't require a hundred lines of code.
- The Rust library is able to through errors as well, even though we will have to figure out how to handle localization. But this issue needs to be solved regardless at some point.